### PR TITLE
Feat:Add RAG Benchmark method

### PR DESCRIPTION
**Pull request by YangQianli92**
**Features**
- New MetaGPT-RAG assessment module, involving RougL, Bleu, Recall, Hit Rate, MRR and other assessment indicators.
- Feel free to review the effects of the different modules of RAG.
- Support customized evaluation dataset, please follow the sample provided by us to modify the structure can be.
- Added Reranker support for Cohere, FlagEmbedding.
- Based on the above work, we have evaluated the various components of MetaGPT, and some of the settings can be viewed in Figure
  - LLM: chatgpt-3.5-1106-turbo
  - Embedding: text-embedding-3-small
  - chunk_size: 256
  - chunk_overleap: 0
  - similarity_top_k: 5
  - ranker_top_n: 3
图片1
image

### CODE REVIEW

File: metagpt/rag/factories/ranker.py [hunk 1]
```diff
@@ -3,14 +3,18 @@
 from llama_index.core.llms import LLM
 from llama_index.core.postprocessor import LLMRerank
 from llama_index.core.postprocessor.types import BaseNodePostprocessor
-
+from llama_index.postprocessor.colbert_rerank import ColbertRerank
+from llama_index.postprocessor.cohere_rerank import CohereRerank
+from llama_index.postprocessor.flag_embedding_reranker import FlagEmbeddingReranker
```
Comment (better629) [inline metagpt/rag/factories/ranker.py:8]: if needs dependence pkgs, put them inside `extras_require.rag` in `setup.py`

File: metagpt/rag/factories/ranker.py [hunk 2]
```diff
@@ -3,14 +3,18 @@
 from llama_index.core.llms import LLM
 from llama_index.core.postprocessor import LLMRerank
 from llama_index.core.postprocessor.types import BaseNodePostprocessor
-
+from llama_index.postprocessor.colbert_rerank import ColbertRerank
```
Comment (better629) [inline metagpt/rag/factories/ranker.py:6]: here use lazy-installation like `https://github.com/geekan/MetaGPT/blob/main/metagpt/rag/factories/ranker.py#L40-L45`, no need to import it here.  Make user to install extra independences manually due to the pkgs are usually large and no need to install when not used.

File: metagpt/rag/schema.py [hunk 3]
```diff
@@ -119,6 +119,18 @@ class ColbertRerankConfig(BaseRankerConfig):
     keep_retrieval_score: bool = Field(default=False, description="Whether to keep the retrieval score in metadata.")
 
 
+class CohereRerankConfig(BaseRankerConfig):
+    model: str = Field(default="rerank-english-v3.0")
+    top_n: int = Field(default=2)
```
Comment (better629) [inline metagpt/rag/schema.py:124]: already has `top_n` in `BaseRankerConfig`

File: metagpt/rag/factories/ranker.py [hunk 4]
```diff
@@ -48,6 +54,13 @@ def _create_colbert_ranker(self, config: ColbertRerankConfig, **kwargs) -> LLMRe
     def _create_object_ranker(self, config: ObjectRankerConfig, **kwargs) -> LLMRerank:
         return ObjectSortPostprocessor(**config.model_dump())
 
+    def _create_cohere_rerank(self, config: CohereRerankConfig, **kwargs) -> LLMRerank:
+        return CohereRerank(**config.model_dump())
+
+    def _create_flag_rerank(self, config: FlagEmbeddingReranker, **kwargs) -> LLMRerank:
```
Comment (better629) [inline metagpt/rag/factories/ranker.py:60]: use lazy-installation like `ColbertRerank`

File: metagpt/rag/benchmark/base.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            )
+        return dataset_config
+
+
+if __name__ == "__main__":
+    benchmark = RAGBenchMark()
+    answer = "是的，根据提供的信息，2023年7月20日，应急管理部和财政部确实联合发布了《因灾倒塌、损坏住房恢复重建救助工作规范》的通知。这份《规范》旨在进一步规范因灾倒塌、损坏住房的恢复重建救助相关工作。它明确了地方各级政府负责实施救助工作，应急管理部和财政部则负责统筹指导。地方财政应安排足够的资金，中央财政也会提供适当的补助。救助资金将通过专账管理，并采取特定的管理方式。救助对象是那些因自然灾害导致住房倒塌或损坏，并向政府提出申请且符合条件的受灾家庭。相关部门将组织调查统计救助对象信息，并建立档案。此外，《规范》还强调了资金发放的具体方式和公开透明的要求。"
+    ground_truth = "“启明行动”是为了防控儿童青少年的近视问题，并发布了《防控儿童青少年近视核心知识十条》。"
+    bleu_avg, bleu1, bleu2, bleu3, bleu4 = benchmark.bleu_score(answer, ground_truth)
+    logger.info(f"bleu_avg = {bleu_avg}")
```
Comment (better629) [inline metagpt/rag/benchmark/base.py:124]: maybe you can use logger.info once to log different metrics, not need to repeat.

File: metagpt/rag/benchmark/base.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    @staticmethod
+    def load_dataset(ds_names: list[str] = ["CRUD"]):
+        with open(os.path.join(EXAMPLE_BENCHMARK_PATH, "dataset_info.json"), "r", encoding="utf-8") as f:
+            infos = json.load(f)
+            dataset_config = DatasetConfig(
+                datasets=[
+                    DatasetInfo(
+                        name=name,
+                        document_files=[
+                            os.path.join(EXAMPLE_BENCHMARK_PATH, name, file)
```
Comment (better629) [inline metagpt/rag/benchmark/base.py:104]: use pathlib.Path not os.path

File: examples/rag_bm.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+         'ground_truth_text':
+         '“启明行动”是为了防控儿童青少年的近视问题，并发布了《防控儿童青少年近视核心知识十条》。'
+            }
+         }
+        """
+        if print_title:
+            self._print_title("RAG Pipeline")
+        try:
+            nodes = await self.engine.aretrieve(question)
+            #    pdb.set_trace()
```
Comment (better629) [inline examples/rag_bm.py:118]: del it

File: metagpt/rag/benchmark/base.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    name: str
+    document_files: List[str]
+    gt_info: List[dict]
+
+
+class DatasetConfig(BaseModel):
+    datasets: List[DatasetInfo]
+
+
+class RAGBenchMark:
```
Comment (better629) [inline metagpt/rag/benchmark/base.py:28]: RAGBenchmark

File: examples/rag_bm.py [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+            self._print_result(nodes, state="Retrieve")
+
+            answer = await self.engine.aquery(question)
+            self._print_result(answer, state="Query")
+
+        except Exception as e:
+            print(e)
+            return {
+                "metrics": {
```
Comment (better629) [inline examples/rag_bm.py:127]: repeat with Line 196-214, can you make it a single method like `RAGBenchmark._set_metrics` with default value, and return desired result if pass different values.
and why not put it (Line 190-194 and so on) into `RAGBenchmark.compute_metric` so you can get the all metrics with one method calling.

File: examples/rag_bm.py [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+        try:
+            nodes = await self.engine.aretrieve(question)
+            #    pdb.set_trace()
+            self._print_result(nodes, state="Retrieve")
+
+            answer = await self.engine.aquery(question)
+            self._print_result(answer, state="Query")
+
+        except Exception as e:
+            print(e)
```
Comment (better629) [inline examples/rag_bm.py:125]: logger

File: metagpt/rag/benchmark/base.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+        result = await self.evaluator.aevaluate(
+            response=response,
+            reference=reference,
+        )
+
+        return result.score
+
+    @staticmethod
+    def load_dataset(ds_names: list[str] = ["CRUD"]):
+        with open(os.path.join(EXAMPLE_BENCHMARK_PATH, "dataset_info.json"), "r", encoding="utf-8") as f:
```
Comment (better629) [inline metagpt/rag/benchmark/base.py:97]: use `from metagpt.utils.common import read_json_file`

File: metagpt/rag/benchmark/base.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+            infos = json.load(f)
+            dataset_config = DatasetConfig(
+                datasets=[
+                    DatasetInfo(
+                        name=name,
+                        document_files=[
+                            os.path.join(EXAMPLE_BENCHMARK_PATH, name, file)
+                            for file in info["document_files"]
+                        ],
+                        gt_info=json.load(
```
Comment (better629) [inline metagpt/rag/benchmark/base.py:107]: `from metagpt.utils.common import read_json_file`

File: examples/rag_bm.py [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+        self.llm = get_rag_llm()
+
+    async def rag_evaluate_pipeline(self, dataset_name: list[str] = ["all"]):
+        dataset_config = self.benchmark.load_dataset(dataset_name)
+    
+        for dataset in dataset_config.datasets:
+            if "all" in dataset_name or dataset.name in dataset_name:
+                output_dir = DATA_PATH / f"{dataset.name}"
+
+                if os.path.exists(output_dir):
```
Comment (better629) [inline examples/rag_bm.py:57]: still have `os.path` and use `output_dir.exists`

File: examples/rag_bm.py [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+                    logger.info("Loading Exists index!")
```
Comment (better629) [inline examples/rag_bm.py:58]: existed

File: examples/rag_bm.py [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+                    logger.info(f"Index Path:{output_dir}")
+                    self.engine = SimpleEngine.from_index(
+                        index_config=FAISSIndexConfig(persist_path=output_dir),
+                        ranker_configs=[ColbertRerankConfig()],
+                        retriever_configs=[FAISSRetrieverConfig(), BM25RetrieverConfig()],
+                    )
+                else:
+                    logger.info("Loading index from document!")
```
Comment (better629) [inline examples/rag_bm.py:66]: documents

File: examples/rag_bm.py [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+        try:
+            nodes = await self.engine.aretrieve(question)
+            self._print_result(nodes, state="Retrieve")
+
+            answer = await self.engine.aquery(question)
+            self._print_result(answer, state="Query")
+
+        except Exception as e:
+            logger.error(e)
+            return self.benchmark._set_metrics(
```
Comment (better629) [inline examples/rag_bm.py:129]: not desired implement. called wrong method.
```
compute_metric
     _set_metrics
```

File: metagpt/rag/benchmark/base.py [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+        log = {
+            "generated_text": generated_text,
+            "ground_truth_text": ground_truth_text,
+            "question": question,
+        }
+
+        return {"metrics": metrics, "log": log}
+
+
+    def bleu_score(self, response: str, reference: str, with_penalty=False) -> float:
```
Comment (better629) [inline metagpt/rag/benchmark/base.py:79]: output type, Union[float, Tuple[float]]

File: metagpt/rag/benchmark/base.py [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+
+        for i, doc in enumerate(reference_docs, start=1):
+            for node in nodes:
+                if node.text in doc:
+                    mrr_sum += 1.0 / i
+                    break
+
+        return mrr_sum / len(reference_docs) if reference_docs else 0.0
+
+    async def SemanticSimilarity(self, response: str, reference: str) -> float:
```
Comment (better629) [inline metagpt/rag/benchmark/base.py:129]: use aa_bb format, same with MRR, HitRate

File: metagpt/rag/schema.py [hunk 19]
```diff
@@ -119,6 +119,16 @@ class ColbertRerankConfig(BaseRankerConfig):
     keep_retrieval_score: bool = Field(default=False, description="Whether to keep the retrieval score in metadata.")
 
 
+class CohereRerankConfig(BaseRankerConfig):
+    model: str = Field(default="rerank-english-v3.0")
+    api_key: str = Field(default="YOUR_COHERE_API")
+
+
+class FlagEmbeddingConfig(BaseRankerConfig):
```
Comment (better629) [inline metagpt/rag/schema.py:127]: why use `xxEmbeddingConfig` not a `xxRankerConfig`

**Comment by better629** [comment]
lgtm

**Review by geekan — APPROVED** [review]

**Comment by YangQianli92** [comment]
In the PR submitted above, there is a slight error in the MRR calculation of the Benchmark metrics, I have submitted another PR for fixing this bug, and all the results are recalculated after fixing the bug!

**Pull request merged**